utils/weighted_random.go

```diff
@@ -0,0 +1,155 @@
+package utils
+
+import (
+	"fmt"
+	"hash/fnv"
+	"math/rand"
+	"reflect"
+	"sort"
+	"sync"
+	"time"
+)
+
+var reflectInt = reflect.TypeOf(int(1))
+var reflectStr = reflect.TypeOf("a")
+var reflectFloat = reflect.TypeOf(float64(1))
+
+type WeightedRandom interface {
+	Get() interface{}
+	GetWithSeed(seed string) (interface{}, error)
+}
+
+type Entry struct {
+	Item   interface{}
+	Weight float32
+}
+
+type internalEntry struct {
+	entry        Entry
+	currentTotal float32
+}
+
+// WeightedRandom selects elements randomly from the list taking into account individual weights.
+// Weight has to be assigned between 0 and 1.
+// Support deterministic results given a particular seed and sortKey
+type weightedRandomImpl struct {
+	entries     []internalEntry
+	totalWeight float32
+	mux         sync.Mutex
+}
+
+func validateEntries(entries []Entry, sortKey string) error {
+	if len(entries) == 0 {
+		return fmt.Errorf("entries is empty")
+	}
+	for _, entry := range entries {
+		if entry.Weight < 0 || entry.Weight > float32(1) {
+			return fmt.Errorf("invalid weight %f", entry.Weight)
+		}
+
+		item := reflect.ValueOf(entry.Item)
```

Comment:
No reflections please.. just add a method on Item interface that returns the sort key

Reply:
done